topology/probes/vpp/vpp.go

```diff
@@ -103,268 +124,508 @@ func (p *Probe) getInterfaceVrfID(ch api.Channel, index uint32) int64 {
 }
 
 func (p *Probe) getInterface(index uint32) *graph.Node {
-	p.Ctx.Graph.RLock()
-	defer p.Ctx.Graph.RUnlock()
-	return p.Ctx.Graph.LookupFirstNode(graph.Metadata{"IfIndex": int64(index), "Type": "vpp"})
+	node, _ := p.intfIndexer.GetNode(int64(index))
+	return node
 }
 
-func (p *Probe) createOrUpdateInterface(ch api.Channel, intf *interfaces.SwInterfaceDetails) *graph.Node {
-	metadata := graph.Metadata{"IfIndex": int64(intf.SwIfIndex), "Type": "vpp"}
+func (p *Probe) createOrUpdateInterface(ch api.Channel, intf *interfaces.SwInterfaceDetails, metadata graph.Metadata) *graph.Node {
+	vrfID := p.getInterfaceVrfID(ch, intf.SwIfIndex)
+	name := interfaceName(intf.InterfaceName)
 
 	p.Ctx.Graph.Lock()
 	defer p.Ctx.Graph.Unlock()
 
+	metadata["Driver"] = "vpp"
+	metadata["Type"] = "interface"
+	metadata["Name"] = name
+	metadata["IfIndex"] = int64(intf.SwIfIndex)
+	metadata["MAC"] = interfaceMAC(intf.L2Address[:intf.L2AddressLength])
+	metadata["MTU"] = int64(intf.LinkMtu)
+	metadata["Speed"] = int64(intf.LinkSpeed)
+	metadata["State"] = interfaceUpDown(intf.AdminUpDown)
+	metadata["Duplex"] = interfaceDuplex(intf.LinkDuplex)
+	metadata["VrfID"] = vrfID
+
+	if strings.HasPrefix(name, "host-") {
+		metadata["InterfaceName"] = name[5:]
+	}
+
 	var err error
-	node := p.Ctx.Graph.LookupFirstNode(metadata)
+	node := p.getInterface(intf.SwIfIndex)
 	if node == nil {
 		node, err = p.Ctx.Graph.NewNode(graph.GenID(), metadata)
 		if err != nil {
 			p.Ctx.Logger.Error(err)
 			return nil
 		}
+		p.NotifyEvent(graph.NodeAdded, node)
```

Comment:
Does NewNode not notified already ?

Reply:
The probe embeds a graph.EventHandler so that listeners can be notified of the events generated by the probe only and not all the events on the graph.